### PR TITLE
Port changes to bjorn

### PR DESCRIPTION
**Pull request by GregBowyer**
This ports changes from my PR into yours.

# Changes
* I fixed a few very small mistakes in tests and the code, nothing major
* I make the integration test build artifacts with `cc` so we lose the checked in `.o` files.
* I changed the API in `Symbol` as `file_offset` is fundimentally useless (we _could_ implement `Seek` on `Archive` but that feels pretty wrong given we can just lookup the index of the entity and provide that.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,3 +7,12 @@ repository = "https://github.com/mdsteele/rust-ar"
...
 readme = "README.md"
+
+[dev-dependencies]
+pretty_assertions = "0.6"
+proptest = "0.10"
+criterion = "0.3"
+itertools = "0.9"
+assert_cmd = "1.0"
+indoc = "0.3"
+tempfile = "3.1"
```
Comment (bjorn3) [inline Cargo.toml:22]: Please add those dependencies in the commits they are used.

File: src/write.rs [hunk 2]
```diff
@@ -150,16 +144,10 @@ impl<W: Write + Seek> Builder<W> {
...
-                        ErrorKind::InvalidInput,
-                        format!(
-                            "More than 4GB of symbol strings??? There are 0x{:x} bytes",
-                            total_symbol_size
-                        ),
-                    )
-                })?,
-            ))?;
+                u32::try_from(total_symbol_size)
+                    .map_err(|_| err!("Too many symbol names `{:#x} bytes`", total_symbol_size))?
```
Comment (bjorn3) [inline src/write.rs]:
```suggestion
                    .map_err(|_| err!("Total symbol name size too much. `{:#x}` bytes", total_symbol_size))?
```

File: src/write.rs [hunk 3]
```diff
@@ -433,14 +408,10 @@ impl<W: Write + Seek> GnuBuilder<W> {
...
-            let msg = format!(
-                "Identifier {:?} was not in the list of \
-                 identifiers passed to GnuBuilder::new()",
-                String::from_utf8_lossy(header.identifier())
-            );
-            return Err(Error::new(ErrorKind::InvalidInput, msg));
-        }
+
+        ensure!(has_name,
+            "Identifier `{}` was not in the list of identifiers passed to GnuBuilder::new()",
```
Comment (bjorn3) [inline src/write.rs]:
```suggestion
            "Identifier {:?} was not in the list of identifiers passed to GnuBuilder::new()",
```

This escapes characters like `\0` and `\n`. Otherwise they would be invisible.

File: src/lib.rs [hunk 4]
```diff
@@ -200,5 +200,51 @@ impl Header {
 
     /// Sets the length of the file, in bytes.
     pub fn set_size(&mut self, size: u64) { self.size = size; }
+
+    /// Validates the header is somewhat sane against the specification
```
Comment (bjorn3) [inline src/lib.rs]: nit:

```suggestion
    /// Validates the header is somewhat sane against the specification.
```

File: src/lib.rs [hunk 5]
```diff
@@ -200,5 +200,51 @@ impl Header {
...
+    pub fn validate(&self) -> std::io::Result<()> {
+        ensure!(num_digits(self.mtime as f64, 10) <= 12, "MTime `{}` > 12 digits", self.mtime);
```
Comment (bjorn3) [inline src/lib.rs]: Why is the `as f64` necessary?

File: src/lib.rs [hunk 6]
```diff
@@ -139,7 +139,7 @@ impl Header {
             mtime: 0,
             uid: 0,
             gid: 0,
-            mode: 0,
+            mode: 0o644,
```
Comment (bjorn3) [inline src/lib.rs:144]: Yay, I can finally read files extracted using `ar` without using `chmod` when using `Header::new` during writing!

File: src/test_support.rs [hunk 7]
```diff
@@ -0,0 +1,45 @@
+//! Test utils for rust-ar, typically `Arbitrary` implementations of things
```
Comment (bjorn3) [inline src/test_support.rs]:
```suggestion
//! Test utils for rust-ar, mostly `Arbitrary` implementations of things.
```

File: src/test_support.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+use proptest::{
+    arbitrary::Arbitrary,
+    collection::vec,
+    prelude::*
+};
+
+/// Arbitrary implementation of Header and Data for feeding into tests
+#[derive(Debug, Clone)]
+pub struct HeaderAndData {
+    /// The header generated for this case
```
Comment (bjorn3) [inline src/test_support.rs]:
```suggestion
    /// The header generated for this case.
```

File: src/test_support.rs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    pub header: Header,
+    /// The data for this case
```
Comment (bjorn3) [inline src/test_support.rs]:
```suggestion
    /// The data for this case.
```

File: src/test_support.rs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    pub data: Vec<u8>,
+}
+
+impl Arbitrary for HeaderAndData {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    /// Produces arbitrary archives for testing
```
Comment (bjorn3) [inline src/test_support.rs]:
```suggestion
    /// Produces arbitrary archives for testing.
```

File: src/test_support.rs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    ///
+    /// **Note:** very tecnically its possible to have `/` and spaces in archive names for
```
Comment (bjorn3) [inline src/test_support.rs]:
```suggestion
    /// **Note:** very technically its possible to have `/` and spaces in archive names for
```

File: src/read.rs [hunk 12]
```diff
@@ -595,10 +597,11 @@ impl<R: Read + Seek> Archive<R> {
                         {
                             str_end += 1;
                         }
+                        let entry_index = lookup_entry_from_offset(&self.entry_headers, file_offset as u64)?;
```
Comment (bjorn3) [inline src/read.rs]:
```suggestion
                        let entry_index = lookup_entry_from_offset(&self.entry_headers, u64::from(file_offset))?;
```

File: tests/gnu_ar.rs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+};
+use itertools::Itertools;
+use anyhow::{Result, anyhow};
+
+#[inline]
+fn as_str(raw: &[u8]) -> String {
+    String::from_utf8_lossy(raw).to_string()
+}
+
+/// Dumb naive function that returns the command if it exists and can pass some basic checks
```
Comment (bjorn3) [inline tests/gnu_ar.rs:19]:
```suggestion
/// Dumb naive function that returns the command if it exists and can pass some basic checks.
```

File: tests/gnu_ar.rs [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+fn command(name: &str) -> Option<Command> {
+    match Command::new(name).arg("--version").output() {
+        Ok(out) => {
+            if out.status.success() && as_str(&out.stdout).contains("GNU") {
+                let mut command = Command::new(name);
+                command.env("TZ", "UTC");
+                Some(command)
+            } else {
+                println!(
+                    "None GNU tool ?\n{}{}",
```
Comment (bjorn3) [inline tests/gnu_ar.rs]:
```suggestion
                    "Non-GNU tool?\nstdout:\n{}\nstderr:\n{}",
```

File: tests/gnu_ar.rs [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+                name, e
+            );
+            None
+        }
+    }
+}
+
+fn fixture_path(name: &str) -> PathBuf {
+    let root_dir =
+        env::var("CARGO_MANIFEST_DIR").expect("$CARGO_MANIFEST_DIR");
```
Comment (bjorn3) [inline tests/gnu_ar.rs]:
```suggestion
    let root_dir = env!("CARGO_MANIFEST_DIR");
```

This way running the test directly also works.

File: Cargo.toml [hunk 16]
```diff
@@ -19,3 +19,6 @@ indoc = "0.3"
 tempfile = "3.1"
 rand = "0.7"
 rand_pcg = "0.2"
+anyhow = "1.0"
```
Comment (bjorn3) [inline Cargo.toml]: I think anyhow is overkill for a test. You could for example return `Result<_, String>` or `Result<_, Box<dyn Error>>`.

File: tests/gnu_ar.rs [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+                .target(target.target_triple)
+                .opt_level(2)
+                .host(target.target_triple)
+                .out_dir(&dir)
+                .file(fixture_path("object_normal_syms.c"))
+                .file(fixture_path("object_weak_syms.c"))
+                .include(fixture_path(""))
+                .try_compile("libnative.a")
+                .map_err(|e| e.into())
+                .and(Ok(dir))
```
Comment (bjorn3) [inline tests/gnu_ar.rs:78]: Maybe run `ar` directly?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: tests/gnu_ar.rs (hunk 17 above)
Comment (GregBowyer) [reply tests/gnu_ar.rs:78]: `cc` doesnt offer much for splitting up the stages so I am not sure how we would want to do that.

Comment (bjorn3) [reply tests/gnu_ar.rs:78]: I mean `Command::new("ar")`. That should work in most cases I think.

File: src/lib.rs (hunk 5 above)
Comment (GregBowyer) [reply src/lib.rs]: there is no conversion from `u64` to `f64`

its ugly but would you prefer this

```rust
#[inline]
fn num_digits<N: Into<u64>>(val: N, radix: u64) -> u32 {
    let mut val = val.into();
    if val == 0 {
        return 1;
    }

    let mut digits = 0;
    while val != 0 {
        val /= radix;
        digits += 1;
    }
    digits
}
```

File: src/write.rs (hunk 3 above)
Comment (GregBowyer) [reply src/write.rs]: Well tbh if thats the intent we should just print the bytes and not try to use a string.

The end user _gave us bytes_ in the first place

File: tests/gnu_ar.rs (hunk 17 above)
Comment (GregBowyer) [reply tests/gnu_ar.rs:78]: Do you mean to extract the just compiled objects?

I think its better to use our own code now as it gives us a test of reading ar files made with a toolchain no?

Comment (bjorn3) [reply tests/gnu_ar.rs:78]: I meant using `ar` directly for creating an archive.

File: src/test_support.rs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    /// different permutations of BSD and GNU archives. We do not test these because the corner case
+    /// is very subtle and most tools do not deal well with archives that have such strange names.
+    fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
+        (
+            vec(any::<u8>(), 0..65 * 1024), // Data
+            r#"[\PC&&[^/[:blank:]]]{1,50}"#, // Ident
+            0..999999999999u64, // Mtime
+            0..999999u32, // Uid
+            0..999999u32, // Gid
+            0..342391u32, // Mode
```
Comment (bjorn3) [inline src/test_support.rs]:
```suggestion
            0..0o7777_7777u32, // Mode
```
